### PR TITLE
Product gallery/swiper 2

### LINKED ISSUE #43255: Product Gallery: Consider implementing pseudo infinite loop effect
### Describe the solution you'd like

Currently, when you scroll to the end of available product thumbnails the next button doesn't do anything.

AFAIK it's not possible to utilize a true infinite loop effect with a pure CSS slider like the Product Gallery block utilizes, but we should be able to scroll to the first thumbnail, like here: https://codepen.io/alvarotrigo/pen/bGqgqeK

### PR DESCRIPTION
**Pull request by roykho**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/43255

This PR addresses the issue of not having infinite scroll on the main image by introducing a library called [Swiper](https://swiperjs.com/) which has many awesome features including touch gestures. Another good feature about Swiper is that you only include the modules you want to use making it lite weight.

Note 1: I updated the code for the pager to show all the images in the thumbnail instead of limited to the setting in the inspector controls as it seems odd.

Note 2: I reverted https://github.com/woocommerce/woocommerce/pull/53318

Note 3: Nothing is done with the dialog box yet.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

> [!NOTE]  
> **Pre-requisite**
> - Make sure to do a `pnpm install && pnpm build` to install the Swiper library.

1. Go to the Appearance->Editor->Templates->Single product template.
2. Replace the Product Image Gallery block with the Product Gallery (beta) block.
3. Save and go to the frontend shop.
4. Choose a product with multiple gallery images and try scrolling through the images and when it reaches the end, you can keep scrolling.
5. Test step 4 with the arrow buttons and touch gestures.
6. Try also clicking on the thumbnail gallery images and ensure they work to change the image in the main area.
7. Try also clicking on the pager dots and ensure they're also working to change the image in the main area.
8. Try clicking on the main image to open the dialog. Ensure the image shown is the image you clicked on.
9. Try navigating to other images via the keyboard arrows and also by clicking on the left/right arrows.
10. Ensure the gallery is infinite loop.
11. Try touch gestures in the dialog and ensure you can swipe to each image.
12. Try testing a product with variation images that also exists as gallery images.
13. When you change the variation, ensure you see the variation image show in the main image area.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/blocks/product-gallery/frontend.tsx [hunk 1]
```diff
@@ -27,84 +26,108 @@ const getContext = ( ns?: string ) =>
...
- * Function returns images per current context.
- */
-const getCurrentImages = ( context: ProductGalleryContext ) => {
-	const { isDialogOpen } = context;
-	return context[
-		isDialogOpen ? 'dialogVisibleImagesIds' : 'visibleImagesIds'
-	];
-};
+const swiperMainContainer = document.querySelector(
+	'.wc-block-product-gallery-large-image'
```
Comment (kmanijak) [inline plugins/woocommerce-blocks/assets/js/blocks/product-gallery/frontend.tsx:30]: Nitpick: we could be more specific with these selectors at least by introducing `id`s to the markup. Targeting by class always feels a bit risky 😅 
If you agree that should be applied to each `querySelector` and `querySelectorAll` (probably exchanging with `getElementById`).

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductGalleryPager.php [hunk 2]
```diff
@@ -86,33 +79,23 @@ protected function render( $attributes, $content, $block ) {
...
 	 * @return string|null The rendered pager HTML, or null if the pager should not be displayed.
 	 */
-	private function render_pager( $product_gallery_images_ids, $pager_display_mode, $number_of_thumbnails ) {
-		if ( $number_of_thumbnails < 2 || 'off' === $pager_display_mode ) {
-			return null;
-		}
-
-		return $this->render_pager_pages( $product_gallery_images_ids, $number_of_thumbnails, $pager_display_mode );
+	private function render_pager( $product_gallery_images_ids, $pager_display_mode ) {
+		return $this->render_pager_pages( $product_gallery_images_ids, $pager_display_mode );
```
Comment (kmanijak) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductGalleryPager.php:85]: I think we should bring the condition back:
- if there's just one image, what's the point of rendering pager?

image

- if pages is "off" it shouldn't be rendered either.

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductGalleryThumbnails.php [hunk 3]
```diff
@@ -48,9 +48,9 @@ protected function get_block_type_uses_context() {
 	 * @return string
 	 */
 	protected function generate_view_all_html( $remaining_thumbnails_count ) {
-		$view_all_html = '<div class="wc-block-product-gallery-thumbnails__thumbnail__overlay" data-wc-on--click="actions.openDialog" data-wc-on--keydown="actions.onViewAllImagesKeyDown" tabindex="0">
-			<span class="wc-block-product-gallery-thumbnails__thumbnail__remaining-thumbnails-count">+%1$s</span>
-			<span class="wc-block-product-gallery-thumbnails__thumbnail__view-all">%2$s</span>
+		$view_all_html = '<div class="wc-block-product-gallery-thumbnails__thumbnail__overlay wc-block-product-gallery-dialog-on-click" data-wc-on--click="actions.openDialog" data-wc-on--keydown="actions.onViewAllImagesKeyDown" tabindex="0">
+			<span class="wc-block-product-gallery-thumbnails__thumbnail__remaining-thumbnails-count wc-block-product-gallery-dialog-on-click">+%1$s</span>
+			<span class="wc-block-product-gallery-thumbnails__thumbnail__view-all wc-block-product-gallery-dialog-on-click">%2$s</span>
```
Comment (kmanijak) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductGalleryThumbnails.php:53]: Isn't it enough to put it on wrapping div? What's the purpose of `wc-block-product-gallery-dialog-on-click` class anyway? I can't see any use of it 🤔 

```suggestion
		$view_all_html = '<div class="wc-block-product-gallery-thumbnails__thumbnail__overlay wc-block-product-gallery-dialog-on-click" data-wc-on--click="actions.openDialog" data-wc-on--keydown="actions.onViewAllImagesKeyDown" tabindex="0">
			<span class="wc-block-product-gallery-thumbnails__thumbnail__remaining-thumbnails-count">+%1$s</span>
			<span class="wc-block-product-gallery-thumbnails__thumbnail__view-all">%2$s</span>
```

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductGalleryLargeImage.php [hunk 4]
```diff
@@ -118,14 +118,14 @@ function ( $carry, $key ) use ( $directives ) {
...
-			'data-wc-watch'          => 'callbacks.scrollInto',
+			'data-wc-bind--hidden'   => '!state.isSelected',
 			'data-wc-bind--tabindex' => 'state.thumbnailTabIndex',
 			'data-wc-on--keydown'    => 'actions.onSelectedLargeImageKeyDown',
 			'data-wc-class--wc-block-woocommerce-product-gallery-large-image__image--active-image-slide' => 'state.isSelected',
 		);
 
 		if ( $context['fullScreenOnClick'] ) {
-			$attributes['class'] .= ' wc-block-woocommerce-product-gallery-large-image__image--full-screen-on-click';
+			$attributes['class'] .= ' wc-block-woocommerce-product-gallery-large-image__image--full-screen-on-click wc-block-product-gallery-dialog-on-click';
```
Comment (kmanijak) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductGalleryLargeImage.php:128]: What's the purpose of `wc-block-product-gallery-dialog-on-click` class?

File: plugins/woocommerce-blocks/assets/js/blocks/product-gallery/frontend.tsx [hunk 5]
```diff
@@ -27,84 +26,108 @@ const getContext = ( ns?: string ) =>
...
+		?.querySelector(
+			`.swiper-slide[data-swiper-slide-index="${ index }"] img`
+		)
+		?.getAttribute( 'data-wc-context' );
+
+	const imageId = imageIdContext
+		? JSON.parse( imageIdContext ).imageId
+		: null;
+
+	return imageId;
```
Comment (kmanijak) [inline plugins/woocommerce-blocks/assets/js/blocks/product-gallery/frontend.tsx:77]: I must say mixing iAPI directives (`data-wc-context`), DOM API (`querySelector` etc) and external JS library doesn't feel right. There's nothing particularly harmful in this piece of code but it clearly shows we're mixing techniques/approaches.

I think we should either migrate fully from iAPI to custom JS library OR rely on iAPI. This comes from iAPI being based on preact and preact doesn't act good with direct DOM manipulations (by direct, I mean[ using native DOM API](https://github.com/nolimits4web/swiper/blob/master/src/core/slide/slideTo.mjs#L119-L152)).

File: plugins/woocommerce-blocks/assets/js/blocks/product-gallery/frontend.tsx [hunk 6]
```diff
@@ -27,84 +26,108 @@ const getContext = ( ns?: string ) =>
...
-		context.selectedImage
-	);
+const setPager = ( imageId: string ) => {
+	(
+		document.querySelector(
+			".wc-block-product-gallery-pager__pager-item[data-wc-context='" +
+				JSON.stringify( { imageId } ) +
+				"']"
+		) as HTMLButtonElement
+	 )?.click();
```
Comment (kmanijak) [inline plugins/woocommerce-blocks/assets/js/blocks/product-gallery/frontend.tsx:87]: I consider it one of the biggest red flags for this solution at the current stage. If we have to mimic users behaviours (DOM element click) to synchronise state of blocks, it means we're doing it wrong. Of course solution may work but it introduces technical debt from a start, increases maintenance cost and further development may be getting harder.

**Review by kmanijak — COMMENTED** [review]
I'm not convinced the external library is the way to go especially in such a setup: mixing iAPI with it.

I'm strongly convinced it has to be _either, or_ and I wouldn't merge the PR in such a state. I tested it and it works fine but it just doesn't feel right to go with that hybrid solution!

What's your take on it? Do you think it's the way to go or you treat it as a POC? I'd love to know your opinion! 🙌

**Comment by roykho** [comment]
@kmanijak thanks for the review and after having some thoughts overnight, I think we shouldn't move forward with this solution for the reasons you mentioned. We could probably do this all via iAPI but would just be a much bigger effort but I think doable. I think it might worth exploring for next cycle.

**Pull request closed without merging**